web/onos-gui/src/app/onos-topo/topodevice.service.ts

```diff
@@ -260,6 +264,71 @@ export class TopoDeviceService {
         );
     }
 
+    watchTopoEntity(errorCb: (e: grpcWeb.Error) => void, updateCb?: (type: Update.Type, entity: Entity) => void) {
+        this.topoDevicesSub = this.onosTopoDeviceService.requestListTopo().pipe(
+            filter(x => x.getUpdate().getObject().getType() === 1)
+        ).subscribe(
+            (resp: SubscribeResponse) => {
+                const name = resp.getUpdate().getObject().getType() + ' ' + resp.getUpdate().getObject().getId();
+                console.log('List Topo Entity response ', name);
+                if (!this.entityList.has(name) &&
+                    (resp.getUpdate().getType() === Update.Type.INSERT || resp.getUpdate().getType() === Update.Type.UNSPECIFIED)) {
+                    const added = this.addTopoEntity(resp.getUpdate().getObject());
+                    if (added && updateCb !== undefined) {
+                        updateCb(resp.getUpdate().getType(), resp.getUpdate().getObject().getEntity());
+                    }
+                } else if (this.entityList.has(name) && resp.getUpdate().getType() === Update.Type.DELETE) {
+                    const removed = this.removeEntity(resp.getUpdate().getObject().getId());
+                    if (removed && updateCb) {
+                        updateCb(resp.getUpdate().getType(), resp.getUpdate().getObject().getEntity());
+                    }
+                } else if (resp.getUpdate().getType() === Update.Type.MODIFY) {
+                    const updated = resp.getUpdate().getObject().getEntity();
+                    this.entityList.set(name, updated);
+                    // const protosHandled: Protocol[] = [];
+                    // this.entityList.get(name).getProtocolsList().forEach((protocol) => {
+                    //     protosHandled.push(protocol.getProtocol());
+                    //     const newProtoState =
+                    //         updated.getProtocolsList().find((p) => p.getProtocol() === protocol.getProtocol());
+                    //     if (protocol.getChannelstate() !== newProtoState.getChannelstate() ||
+                    //         protocol.getConnectivitystate() !== newProtoState.getConnectivitystate() ||
+                    //         protocol.getServicestate() !== newProtoState.getServicestate()) {
+                    //         this.entityList.set(name, resp.getUpdate().getObject().getEntity());
+                    //     }
+                    // });
+                    // In case there are some in the new set that are not in the old - just push them
+                    // updated.getProtocolsList().filter((p) => !protosHandled.includes(p.getProtocol())).forEach((p) => {
+                    //     this.entityList.set(name, resp.getUpdate().getObject().getEntity());
+                    // });
+                } else {
+                    console.log('Unhandled Topo update', resp.getUpdate().getType(), resp.getUpdate().getObject().getId());
```

Comment:
maybe call the `errorCb` here

Reply:
okay will do, I am also adding a stopWatchingTopoEntity function